astropy/io/fits/src/compressionmodule.c

```diff
@@ -1035,11 +1169,18 @@ PyObject* compression_decompress_hdu(PyObject* self, PyObject* args)
 
     /* Create and allocate a new array for the decompressed data */
     outdata = (PyArrayObject*) PyArray_SimpleNew(zndim, znaxis, npdatatype);
+    if (outdata == NULL) {
+        goto fail;
+    }
 
     fits_read_img(fileptr, datatype, 1, arrsize, NULL, PyArray_DATA(outdata),
                   &anynul, &status);
+    /* At this point we need to let CFITSIO clean up the tableptr and the
+       compressed tile cache. */
+    free_columns_manually = 0;
     if (status != 0) {
         process_status_err(status);
+        Py_DECREF(outdata);
```

Comment:
I'm not really sure about that, I think that not decrementing the reference count would leak the `PyArrayObject` but the data has also been passed to `fits_read_img`, which means that the data of the array could be shared and cleaned up when one closes the file.